stdlib/public/Concurrency/AsyncStreamBuffer.swift

```diff
@@ -120,57 +120,32 @@ extension AsyncStream {
     }
 
     func yield(_ value: __owned Element) -> Continuation.YieldResult {
-      var result: Continuation.YieldResult
+      let result: Continuation.YieldResult
       lock()
       let limit = state.limit
       let count = state.pending.count
 
       if !state.continuations.isEmpty {
+        // Presence of continuations implies no pending elements.
+        // TODO: which assertion flavor should be used?
+        assert(
+          state.pending.isEmpty,
+          "Continuations should imply no pending elements."
+        )
         let continuation = state.continuations.removeFirst()
-        if count > 0 {
-          if !state.terminal {
-            switch limit {
-            case .unbounded:
-              state.pending.append(value)
-              result = .enqueued(remaining: .max)
-            case .bufferingOldest(let limit):
-              if count < limit {
-                state.pending.append(value)
-                result = .enqueued(remaining: limit - (count + 1))
-              } else {
-                result = .dropped(value)
-              }
-            case .bufferingNewest(let limit):
-              if count < limit {
-                state.pending.append(value)
-                result = .enqueued(remaining: limit - (count + 1))
-              } else if count > 0 {
-                result = .dropped(state.pending.removeFirst())
-                state.pending.append(value)
-              } else {
-                result = .dropped(value)
-              }
-            }
-          } else {
-            result = .terminated
-          }
-          let toSend = state.pending.removeFirst()
-          unlock()
-          continuation.resume(returning: toSend)
-        } else if state.terminal {
+        if state.terminal {
```

Comment:
per @FranzBusch's prior [observation](https://github.com/swiftlang/swift/pull/75878#discussion_r1716924634), the branches in which we were handling cases that implied both pending elements and outstanding continuations should in fact be dead. the rationale is that elements should only be buffered if there are no awaiting consumers, and continuations are stored only if there are no pending elements. sanity checks appreciated regarding this analysis, however.